### ISSUE TITLE
V2 : support URL in require()

### ISSUE THREAD

**Issue by z3dev** [jscad/OpenJSCAD.org#671]
Comments from #245 

Looking at the docs, include via URL isn't possible yet... Would it be possible to add it? It'd be amazingly handy for referencing a file on GitHub, Gist or similar. Obviously cross-origin could be a pain, but I think GitHub at least is ok from that point of view?

Any feature that allows either of the 2 options below is welcomed!
> specifying more than one JS file in a URL (one file contains the main() function, other are helper libraries)
> specifying one file in the URL, but that file can require() from other files in the same location

This would finally allow giving out URLs of objects you have made that uses lower level libraries, without requiring you to copy/paste the libraries into a single file, or forcing the use of node.js for the same task.

## Steps to Reproduce the Problem

  1.
  2.
  3.

## Specifications

  - Version:
  - Platform:
  - Environment: (browser, local server, node.js etc)

**Comment by SimonClark** [comment]
I agree that this would be a powerful capability to add.  I think to drive adoption of jscad means to build a community.  Doing that means making sharing and discovering of jscad models and assets dead-simple, and empowering the community to develop mix-ins.

This seems like a better way enable some of that than anything I've come up with :)

Do we have a sense of how complicated it's likely to be.  We need a custom version of require(), I guess

**Comment by z3dev** [comment]
Thinking a little further... over-loading 'require()' could be dangerous. If hacked then designs will be forever forced to use require(), and there's ES6 'import()' on the horizon.

Also, there's the whole issue of handling heiarchies of require() statements, both normal and URL based. UG.

We should allow advanced users to use more functionality within designs, such as promises. This would make almost everything possible.

**Comment by z3dev** [comment]
@SimonClark I have some hacking good news. I was able to change the core library to handle normal or async versions of main functions. This means that users can do most anything now, including network fetches, etc.

I just know that something else will break, so further testing is required.

**Comment by SimonClark** [comment]
Fantastic!  I need to find some time to play with it.  Got lots of ideas

**Comment by hrgdavor** [comment]
@z3dev downloading resources probably should go through different method other than require.
That way whatever direction goes for require/import for javascript dependencies, it does not have to break resource fetching.
And this could then make it easier to go in direction of supporting promises form main, and scripts taht use external resources that need to be fetched async.

sync XHR is perfectly ok and allowed inside workers tho.
